### ISSUE TITLE
[JENKINS-49050] I can view my sequential declarative pipeline stages in the pipeline visualization plugin

### ISSUE THREAD

**Issue by jenkins-infra-bot** [jenkinsci/blueocean-plugin#4648]
<p>For Declarative 1.3 (<a href="https://issues.jenkins.io/browse/JENKINS-46809" title="Allow sequential stages inside parallel in Declarative syntax" class="issue-link" data-issue-key="JENKINS-46809"><del>JENKINS-46809</del></a>), we're adding support for sequential groups of stages, either as part of a set of parallel stages or on its own in a top-level stage. I won't bother going into all the rationale behind it. =) See <a href="https://github.com/jenkinsci/pipeline-model-definition-plugin/pull/227" class="external-link" target="_blank" rel="nofollow noopener">the Declarative PR</a> for the relevant changes, examples, discussion, etc. </p>

<p>This will require changes for both visualization and the editor:</p>
<ul>
	<li><em>Visualization</em> - see my pathetic attempts at mockups attached here. <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jamesdumay" class="user-hover" rel="jamesdumay">jamesdumay</a> probably will come up with something better in like five seconds. =)</li>
	<li><em>Editor</em> - Should be something like we have for parallel stages - reordering would be good too.</li>
</ul>

---
<details><summary><i>Originally reported by <img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/abayer.png" title="abayer's avatar" /> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=abayer">abayer</a>, imported from: <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-49050" target="_blank">I can view my sequential declarative pipeline stages in the pipeline visualization plugin</a></i></summary>
<i><ul>
<li><b>assignee</b>: <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=nicu">nicu</a>
<li><b>status</b>: Closed
<li><b>priority</b>: Major
<li><b>component(s)</b>: blueocean-plugin
<li><b>resolution</b>: Fixed
<li><b>resolved</b>: 2018-08-10T03:44:05+00:00
<li><b>votes</b>: 18
<li><b>watchers</b>: 40
<li><b>imported</b>: 2025-12-01
</ul></i>
<details><summary>Raw content of original issue</summary>

<pre>
<p>For Declarative 1.3 (<a href="https://issues.jenkins.io/browse/JENKINS-46809" title="Allow sequential stages inside parallel in Declarative syntax" class="issue-link" data-issue-key="JENKINS-46809"><del>JENKINS-46809</del></a>), we're adding support for sequential groups of stages, either as part of a set of parallel stages or on its own in a top-level stage. I won't bother going into all the rationale behind it. =) See <a href="https://github.com/jenkinsci/pipeline-model-definition-plugin/pull/227" class="external-link" target="_blank" rel="nofollow noopener">the Declarative PR</a> for the relevant changes, examples, discussion, etc. </p>

<p>This will require changes for both visualization and the editor:</p>
<ul>
	<li><em>Visualization</em> - see my pathetic attempts at mockups attached here. <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jamesdumay" class="user-hover" rel="jamesdumay">jamesdumay</a> probably will come up with something better in like five seconds. =)</li>
	<li><em>Editor</em> - Should be something like we have for parallel stages - reordering would be good too.</li>
</ul>
</pre>
</details>
</details>
<details><summary><i>5 attachments</i></summary>

- [image-2018-08-07-20-14-01-469.png](https://issues.jenkins.io/secure/attachment/43636/image-2018-08-07-20-14-01-469.png)
  > ![image-2018-08-07-20-14-01-469.png](https://issues.jenkins.io/secure/attachment/43636/image-2018-08-07-20-14-01-469.png)
- [image-2018-09-13-14-31-36-029.png](https://issues.jenkins.io/secure/attachment/44453/image-2018-09-13-14-31-36-029.png)
  > ![image-2018-09-13-14-31-36-029.png](https://issues.jenkins.io/secure/attachment/44453/image-2018-09-13-14-31-36-029.png)
- [proof.png](https://issues.jenkins.io/secure/attachment/43751/proof.png)
  > ![proof.png](https://issues.jenkins.io/secure/attachment/43751/proof.png)
- [sequential-group-parallel.png](https://issues.jenkins.io/secure/attachment/41179/sequential-group-parallel.png)
  > ![sequential-group-parallel.png](https://issues.jenkins.io/secure/attachment/41179/sequential-group-parallel.png)
- [sequential-group-top-level.png](https://issues.jenkins.io/secure/attachment/41178/sequential-group-top-level.png)
  > ![sequential-group-top-level.png](https://issues.jenkins.io/secure/attachment/41178/sequential-group-top-level.png)
</details>

**Comment by jenkins-infra-bot** [comment]
- _Remote link associated with this issue:_


  - [#​1784](https://github.com/jenkinsci/blueocean-plugin/pull/1784)

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/michaelneale.png" title="michaelneale's avatar" /> michaelneale:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-49050?focusedId=325940&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-325940">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>hey <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=sophistifunk" class="user-hover" rel="sophistifunk">sophistifunk</a> any guesses on how much work this would be for both vis and editor? </p>
</pre>
</details>

</ul>
</details>

<p>hey <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=sophistifunk" class="user-hover" rel="sophistifunk">sophistifunk</a> any guesses on how much work this would be for both vis and editor? </p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>sophistifunk:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-49050?focusedId=325950&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-325950">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Renderer: zero for multi-stage-parallels, as that's already done. As for "grouping" at the top level, it's mostly about cooking up a sensible visual language rather than layout changes.</p>

<p> </p>

<p>Editor-wise, I'm not sure. We need to replace the editor's mixed-in layout code with the renderer layout class, which was extracted from the renderer for this purpose and should be done either way.</p>

<p> </p>

<p>Also, we would need changes to the `PipelineNodeGraphVisitor` to support it. I have ideas about refactoring that to make it less... intimidating... to change, but I'll need to run them by Vivek and probably a few people on pipeline to make sure I'm not talking out of my ass <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/></p>
</pre>
</details>

</ul>
</details>

<p>Renderer: zero for multi-stage-parallels, as that's already done. As for "grouping" at the top level, it's mostly about cooking up a sensible visual language rather than layout changes.</p>

<p> </p>

<p>Editor-wise, I'm not sure. We need to replace the editor's mixed-in layout code with the renderer layout class, which was extracted from the renderer for this purpose and should be done either way.</p>

<p> </p>

<p>Also, we would need changes to the `PipelineNodeGraphVisitor` to support it. I have ideas about refactoring that to make it less... intimidating... to change, but I'll need to run them by Vivek and probably a few people on pipeline to make sure I'm not talking out of my ass <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/></p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/michaelneale.png" title="michaelneale's avatar" /> michaelneale:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-49050?focusedId=325952&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-325952">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>thanks <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=sophistifunk" class="user-hover" rel="sophistifunk">sophistifunk</a> - so it sounds like "not a small thing" </p>
</pre>
</details>

</ul>
</details>

<p>thanks <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=sophistifunk" class="user-hover" rel="sophistifunk">sophistifunk</a> - so it sounds like "not a small thing" </p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/vivek.png" title="vivek's avatar" /> vivek:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-49050?focusedId=330449&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-330449">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Yeah we need changes in nodes  API, opened this ticket: <a href="https://issues.jenkins-ci.org/browse/JENKINS-49779" class="external-link" target="_blank" rel="nofollow noopener">https://issues.jenkins-ci.org/browse/JENKINS-49779</a>.</p>
</pre>
</details>

</ul>
</details>

<p>Yeah we need changes in nodes  API, opened this ticket: <a href="https://issues.jenkins-ci.org/browse/JENKINS-49779" class="external-link" target="_blank" rel="nofollow noopener">https://issues.jenkins-ci.org/browse/JENKINS-49779</a>.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>horangs:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-49050?focusedId=341138&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-341138">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Hello.
I was looking forward to implementing this sequential in parallel feature. Yesterday finally I can use this in Jenkinsfile as code. But I didn't know visualization does not support yet! </p>

<p>Please assign right person and consider very critical thing to do.</p>
</pre>
</details>

</ul>
</details>

<p>Hello.<br/>
I was looking forward to implementing this sequential in parallel feature. Yesterday finally I can use this in Jenkinsfile as code. But I didn't know visualization does not support yet! </p>

<p>Please assign right person and consider very critical thing to do.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>jbriden:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-49050?focusedId=341342&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-341342">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>We will be focusing first on visualization of sequential stages within the UI. We've begun that work. Editing needs significant review so that we can create an intuitive user experience. In the sake of hyper focusing on delivering high quality experiences for the community very quickly, we are taking one UX task at a time. Stay tuned! <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/></p>
</pre>
</details>

</ul>
</details>

<p>We will be focusing first on visualization of sequential stages within the UI. We've begun that work. Editing needs significant review so that we can create an intuitive user experience. In the sake of hyper focusing on delivering high quality experiences for the community very quickly, we are taking one UX task at a time. Stay tuned! <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/></p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/olamy.jpg" title="olamy's avatar" /> olamy:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-49050?focusedId=341624&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-341624">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>first we need to improve backend rest api.</p>
</pre>
</details>

</ul>
</details>

<p>first we need to improve backend rest api.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/olamy.jpg" title="olamy's avatar" /> olamy:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-49050?focusedId=345080&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-345080">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>pr <a href="https://github.com/jenkinsci/blueocean-plugin/pull/1778" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/blueocean-plugin/pull/1778</a></p>
</pre>
</details>

</ul>
</details>

<p>pr <a href="https://github.com/jenkinsci/blueocean-plugin/pull/1778" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/blueocean-plugin/pull/1778</a></p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>sophistifunk:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-49050?focusedId=345085&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-345085">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=vivek" class="user-hover" rel="vivek">vivek</a> I can take this over at the end of the week when I'm done with these last few PoC tasks if you like, let Nicu enjoy his time seeing the sights. The code's reasonably fresh in my mind because I've been using parts of it for the demo.</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=vivek" class="user-hover" rel="vivek">vivek</a> I can take this over at the end of the week when I'm done with these last few PoC tasks if you like, let Nicu enjoy his time seeing the sights. The code's reasonably fresh in my mind because I've been using parts of it for the demo.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/vivek.png" title="vivek's avatar" /> vivek:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-49050?focusedId=345088&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-345088">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=sophistifunk" class="user-hover" rel="sophistifunk">sophistifunk</a> Fine with me,  <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=nicu" class="user-hover" rel="nicu">nicu</a> ok with you? We would rather have you enjoy your vacation<img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/></p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=sophistifunk" class="user-hover" rel="sophistifunk">sophistifunk</a> Fine with me,  <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=nicu" class="user-hover" rel="nicu">nicu</a> ok with you? We would rather have you enjoy your vacation<img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/></p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>nicu:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-49050?focusedId=345090&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-345090">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Thanks <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=sophistifunk" class="user-hover" rel="sophistifunk">sophistifunk</a>. I have all the debugging code already written locally, and I think I have an easy fix for this. I just talked with <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=olamy" class="user-hover" rel="olamy">olamy</a> about a small bug that prevents me from testing my fix, once that is done, I think I can open a PR without much trouble. If I can't finish by the end of the week or it requires too much time I will contact you and hand over what I already have. I will keep everyone updated here with the progress.</p>
</pre>
</details>

</ul>
</details>

<p>Thanks <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=sophistifunk" class="user-hover" rel="sophistifunk">sophistifunk</a>. I have all the debugging code already written locally, and I think I have an easy fix for this. I just talked with <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=olamy" class="user-hover" rel="olamy">olamy</a> about a small bug that prevents me from testing my fix, once that is done, I think I can open a PR without much trouble. If I can't finish by the end of the week or it requires too much time I will contact you and hand over what I already have. I will keep everyone updated here with the progress.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>nicu:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-49050?focusedId=345612&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-345612">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=cbercea" class="user-hover" rel="cbercea">cbercea</a> here is the preview of <a href="https://github.com/NicuPascu/app-store-demo/blob/sequential_parallel_stages/Jenkinsfile" class="external-link" target="_blank" rel="nofollow noopener">this Jenkinsfile</a> that we talked about in yesterdays meeting. Let me know what you think.</p>

<p><span class="image-wrap" style=""><img src="https://issues.jenkins.io/secure/attachment/43636/43636_image-2018-08-07-20-14-01-469.png" style="border: 0px solid black" /></span></p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=cbercea" class="user-hover" rel="cbercea">cbercea</a> here is the preview of <a href="https://github.com/NicuPascu/app-store-demo/blob/sequential_parallel_stages/Jenkinsfile" class="external-link" target="_blank" rel="nofollow noopener">this Jenkinsfile</a> that we talked about in yesterdays meeting. Let me know what you think.</p>

<p><span class="image-wrap" style=""><img src="https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-misc/refs/heads/main/attachments/43636/image-2018-08-07-20-14-01-469.png" style="border: 0px solid black" /></span></p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>stevenfoster:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-49050?focusedId=345618&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-345618">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>From the screenshot, I guess that the stage group name ('multiple-stages' in the example Jenkinsfile) won't be displayed anywhere?</p>
</pre>
</details>

</ul>
</details>

<p>From the screenshot, I guess that the stage group name ('multiple-stages' in the example Jenkinsfile) won't be displayed anywhere?</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/cbercea.png" title="cbercea's avatar" /> cbercea:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-49050?focusedId=345678&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-345678">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=nicu" class="user-hover" rel="nicu">nicu</a> I really like the idea of no longer combining the top most stage name with the name of the "stack". In cases where this happens, what would we name that "parent"? Just curious. Otherwise this looks exactly like what we talked about. Nice!</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=nicu" class="user-hover" rel="nicu">nicu</a> I really like the idea of no longer combining the top most stage name with the name of the "stack". In cases where this happens, what would we name that "parent"? Just curious. Otherwise this looks exactly like what we talked about. Nice!</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>nicu:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-49050?focusedId=345696&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-345696">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=stevenfoster" class="user-hover" rel="stevenfoster">stevenfoster</a> that is correct, the container stage ('multiple-stages') can't have any steps so that is why we decided to not include it anywhere.</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=stevenfoster" class="user-hover" rel="stevenfoster">stevenfoster</a> that is correct, the container stage ('multiple-stages') can't have any steps so that is why we decided to not include it anywhere.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>nicu:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-49050?focusedId=345697&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-345697">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Thanks <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=cbercea" class="user-hover" rel="cbercea">cbercea</a>, as for the edge case that you mention, I don't really understand it, I will PM you later so we can clarify.</p>
</pre>
</details>

</ul>
</details>

<p>Thanks <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=cbercea" class="user-hover" rel="cbercea">cbercea</a>, as for the edge case that you mention, I don't really understand it, I will PM you later so we can clarify.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>nicu:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-49050?focusedId=345700&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-345700">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Opened <a href="https://github.com/jenkinsci/blueocean-plugin/pull/1784" class="external-link" target="_blank" rel="nofollow noopener">PR#1784</a></p>
</pre>
</details>

</ul>
</details>

<p>Opened <a href="https://github.com/jenkinsci/blueocean-plugin/pull/1784" class="external-link" target="_blank" rel="nofollow noopener">PR#1784</a></p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>mgrinberg:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-49050?focusedId=345930&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-345930">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Hi <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=nicu" class="user-hover" rel="nicu">nicu</a>,</p>

<p> </p>

<p>Are you aware in which version of blue-ocean can we expect this feature?</p>

<p> </p>

<p>Thanks for your work on this feature btw.</p>
</pre>
</details>

</ul>
</details>

<p>Hi <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=nicu" class="user-hover" rel="nicu">nicu</a>,</p>

<p> </p>

<p>Are you aware in which version of blue-ocean can we expect this feature?</p>

<p> </p>

<p>Thanks for your work on this feature btw.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>nicu:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-49050?focusedId=346102&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-346102">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Hi <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=mgrinberg" class="user-hover" rel="mgrinberg">mgrinberg</a>, this feature will land in the next version of blueocean, which will be 1.8</p>
</pre>
</details>

</ul>
</details>

<p>Hi <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=mgrinberg" class="user-hover" rel="mgrinberg">mgrinberg</a>, this feature will land in the next version of blueocean, which will be 1.8</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/feliwir.png" title="feliwir's avatar" /> feliwir:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-49050?focusedId=346284&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-346284">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>This new feature in BlueOcean 1.8 did definetly not fix visualization of parallel sequential stages, but instead break it completly: 
 <span class="image-wrap" style=""><a id="43751_thumb" href="https://issues.jenkins.io/secure/attachment/43751/43751_proof.png" title="proof.png" file-preview-type="image" file-preview-id="43751" file-preview-title="proof.png"><img src="https://issues.jenkins.io/secure/thumbnail/43751/_thumb_43751.png" style="border: 0px solid black" role="presentation"/></a></span> </p>

<p>Each of Windows, Linux etc. is one sequential stage. The corresponding Jenkinsfile looks like this:
<a href="https://gist.github.com/feliwir/75864cb641c448446f836056fc698d8a" class="external-link" target="_blank" rel="nofollow noopener">https://gist.github.com/feliwir/75864cb641c448446f836056fc698d8a</a></p>

<p>While the build is running everything is bugging out, things are moving all over the screen, but not where they're supposed to be. Any advice why the visualization is completly bugged now is welcome!</p>
</pre>
</details>

</ul>
</details>

<p>This new feature in BlueOcean 1.8 did definetly not fix visualization of parallel sequential stages, but instead break it completly: <br/>
 <span class="image-wrap" style=""><a id="43751_thumb" href="https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-last/refs/heads/main/attachments/43751/proof.png" title="proof.png" file-preview-type="image" file-preview-id="43751" file-preview-title="proof.png"><img src="https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-last/refs/heads/main/attachments/43751/proof.png" style="border: 0px solid black" role="presentation"/></a></span> </p>

<p>Each of Windows, Linux etc. is one sequential stage. The corresponding Jenkinsfile looks like this:<br/>
<a href="https://gist.github.com/feliwir/75864cb641c448446f836056fc698d8a" class="external-link" target="_blank" rel="nofollow noopener">https://gist.github.com/feliwir/75864cb641c448446f836056fc698d8a</a></p>

<p>While the build is running everything is bugging out, things are moving all over the screen, but not where they're supposed to be. Any advice why the visualization is completly bugged now is welcome!</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>nicu:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-49050?focusedId=346293&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-346293">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Hi <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=feliwir" class="user-hover" rel="feliwir">feliwir</a>, I looked over your Jenkinsfile and saw a problem. Per the <a href="https://jenkins.io/blog/2018/07/02/whats-new-declarative-piepline-13x-sequential-stages/" class="external-link" target="_blank" rel="nofollow noopener">sequential stages documentation</a> each sequential parallel group needs to be contained by a named Stage that doesn't have any steps.</p>

<p>As an example, your sequential stages called 'Build' and 'Test' need to have another container stage with a name that is the immediate child of the 'stages' container.</p>

<p>Please take a look at <a href="https://github.com/NicuPascu/app-store-demo/blob/sequential_parallel_stages/Jenkinsfile" class="external-link" target="_blank" rel="nofollow noopener">this Jenkinsfile</a> for an example.</p>
</pre>
</details>

</ul>
</details>

<p>Hi <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=feliwir" class="user-hover" rel="feliwir">feliwir</a>, I looked over your Jenkinsfile and saw a problem. Per the <a href="https://jenkins.io/blog/2018/07/02/whats-new-declarative-piepline-13x-sequential-stages/" class="external-link" target="_blank" rel="nofollow noopener">sequential stages documentation</a> each sequential parallel group needs to be contained by a named Stage that doesn't have any steps.</p>

<p>As an example, your sequential stages called 'Build' and 'Test' need to have another container stage with a name that is the immediate child of the 'stages' container.</p>

<p>Please take a look at <a href="https://github.com/NicuPascu/app-store-demo/blob/sequential_parallel_stages/Jenkinsfile" class="external-link" target="_blank" rel="nofollow noopener">this Jenkinsfile</a> for an example.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/carelc.svg" title="carelc's avatar" /> carelc:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-49050?focusedId=346296&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-346296">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Will this also work for "Scripted Pipeline syntax" or do one need to start migrating all "scripted" scripts to "declarative"?</p>
</pre>
</details>

</ul>
</details>

<p>Will this also work for "Scripted Pipeline syntax" or do one need to start migrating all "scripted" scripts to "declarative"?</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/olamy.jpg" title="olamy's avatar" /> olamy:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-49050?focusedId=346300&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-346300">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=carelc" class="user-hover" rel="carelc">carelc</a> it's only supported with declarative</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=carelc" class="user-hover" rel="carelc">carelc</a> it's only supported with declarative</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/feliwir.png" title="feliwir's avatar" /> feliwir:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-49050?focusedId=346301&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-346301">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=nicu" class="user-hover" rel="nicu">nicu</a> i did look at the jenkinsfile you've linked above very closely, but i can't spot any differences. The "Build" and "Test" stage are directly in a stages block, just like your sequential stages (<a href="https://github.com/NicuPascu/app-store-demo/blob/sequential_parallel_stages/Jenkinsfile#L19" class="external-link" target="_blank" rel="nofollow noopener">here</a>). The nesting doesn't seem to be different in any way or did i miss sth?</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=nicu" class="user-hover" rel="nicu">nicu</a> i did look at the jenkinsfile you've linked above very closely, but i can't spot any differences. The "Build" and "Test" stage are directly in a stages block, just like your sequential stages (<a href="https://github.com/NicuPascu/app-store-demo/blob/sequential_parallel_stages/Jenkinsfile#L19" class="external-link" target="_blank" rel="nofollow noopener">here</a>). The nesting doesn't seem to be different in any way or did i miss sth?</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/abayer.png" title="abayer's avatar" /> abayer:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-49050?focusedId=346303&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-346303">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Yeah, <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=feliwir" class="user-hover" rel="feliwir">feliwir</a>’s Jenkinsfile looks right to me - here’s how the stages are laid out:</p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">Continuous (children are parallel) -&gt;
  Windows (children are sequential) -&gt;
    Build
    Test
  Linux (children are sequential) -&gt;
    Build
    Test
    Coverage
  MemCheck (children are sequential) -&gt;
    Build
    Memcheck 
  Android (children are sequential) -&gt;
    Build
  iOS (children are sequential) -&gt;
    Build
</pre>
</div></div>

<p>It’s a valid Jenkinsfile and <b>should</b> render correctly. <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=feliwir" class="user-hover" rel="feliwir">feliwir</a>, could you please open a new JIRA for this and include your Jenkinsfile there? Thanks.</p>
</pre>
</details>

</ul>
</details>

<p>Yeah, <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=feliwir" class="user-hover" rel="feliwir">feliwir</a>’s Jenkinsfile looks right to me - here’s how the stages are laid out:</p>

<pre>
Continuous (children are parallel) ->
  Windows (children are sequential) ->
    Build
    Test
  Linux (children are sequential) ->
    Build
    Test
    Coverage
  MemCheck (children are sequential) ->
    Build
    Memcheck 
  Android (children are sequential) ->
    Build
  iOS (children are sequential) ->
    Build
</pre>

<p>It’s a valid Jenkinsfile and <b>should</b> render correctly. <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=feliwir" class="user-hover" rel="feliwir">feliwir</a>, could you please open a new JIRA for this and include your Jenkinsfile there? Thanks.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/feliwir.png" title="feliwir's avatar" /> feliwir:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-49050?focusedId=346306&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-346306">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I've create the new issue: <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=abayer" class="user-hover" rel="abayer">abayer</a> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=nicu" class="user-hover" rel="nicu">nicu</a>
See <a href="https://issues.jenkins-ci.org/browse/JENKINS-53048" class="external-link" target="_blank" rel="nofollow noopener">https://issues.jenkins-ci.org/browse/JENKINS-53048</a></p>
</pre>
</details>

</ul>
</details>

<p>I've create the new issue: <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=abayer" class="user-hover" rel="abayer">abayer</a> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=nicu" class="user-hover" rel="nicu">nicu</a><br/>
See <a href="https://issues.jenkins-ci.org/browse/JENKINS-53048" class="external-link" target="_blank" rel="nofollow noopener">https://issues.jenkins-ci.org/browse/JENKINS-53048</a></p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>hasselmm:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-49050?focusedId=346589&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-346589">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Looks awesome! Thanks a lot!</p>
</pre>
</details>

</ul>
</details>

<p>Looks awesome! Thanks a lot!</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/olamy.jpg" title="olamy's avatar" /> olamy:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-49050?focusedId=349012&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-349012">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=horangs" class="user-hover" rel="horangs">horangs</a> any Jenkinsfile to reproduce that?</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=horangs" class="user-hover" rel="horangs">horangs</a> any Jenkinsfile to reproduce that?</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>horangs:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-49050?focusedId=349013&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-349013">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=nicu" class="user-hover" rel="nicu">nicu</a> and <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=abayer" class="user-hover" rel="abayer">abayer</a></p>

<p> </p>

<p>Today I can test this visualization, and it shows me like below. Nice works.</p>

<p><span class="image-wrap" style=""><img src="https://issues.jenkins.io/secure/attachment/44453/44453_image-2018-09-13-14-31-36-029.png" style="border: 0px solid black" /></span></p>

<p>But there's one big concern.</p>

<p>In my pipeline code,</p>

<p>stage('Parallel') : it really does not have meaning.</p>

<p>stage('Feature A') and stage('Feature B') and stage('Feature C') : now there's no way to see, (it means sequential actions group(type) name)</p>

<p>but this is more important and user should be aware of it.</p>

<p> </p>

<p>How do you think of this? And please let me know related issue number if exists. Or not, can I create new issue for this?</p>

<p> </p>

<p>Thank you.</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=nicu" class="user-hover" rel="nicu">nicu</a> and <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=abayer" class="user-hover" rel="abayer">abayer</a></p>

<p> </p>

<p>Today I can test this visualization, and it shows me like below. Nice works.</p>

<p><span class="image-wrap" style=""><img src="https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-last/refs/heads/main/attachments/44453/image-2018-09-13-14-31-36-029.png" style="border: 0px solid black" /></span></p>

<p>But there's one big concern.</p>

<p>In my pipeline code,</p>

<p>stage('Parallel') : it really does not have meaning.</p>

<p>stage('Feature A') and stage('Feature B') and stage('Feature C') : now there's no way to see, (it means sequential actions group(type) name)</p>

<p>but this is more important and user should be aware of it.</p>

<p> </p>

<p>How do you think of this? And please let me know related issue number if exists. Or not, can I create new issue for this?</p>

<p> </p>

<p>Thank you.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>horangs:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-49050?focusedId=349014&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-349014">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=olamy" class="user-hover" rel="olamy">olamy</a> Oh, please forget it, now I guess there's no problem. Please look at my comment anyway upside.</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=olamy" class="user-hover" rel="olamy">olamy</a> Oh, please forget it, now I guess there's no problem. Please look at my comment anyway upside.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>hasselmm:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-49050?focusedId=349023&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-349023">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=horangs" class="user-hover" rel="horangs">horangs</a> see <a href="https://issues.jenkins.io/browse/JENKINS-53127" title="Parallel sequential stages should display stage label" class="issue-link" data-issue-key="JENKINS-53127"><del>JENKINS-53127</del></a>. Unfortunately nobody triaged that issue yet.</p>

<p>Also much more important: <a href="https://issues.jenkins.io/browse/JENKINS-53547" title="Parallel sequential pipeline rendering appears distorted while jobs are running" class="issue-link" data-issue-key="JENKINS-53547">JENKINS-53547</a></p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=horangs" class="user-hover" rel="horangs">horangs</a> see <a href="https://issues.jenkins.io/browse/JENKINS-53127" title="Parallel sequential stages should display stage label" class="issue-link" data-issue-key="JENKINS-53127"><del>JENKINS-53127</del></a>. Unfortunately nobody triaged that issue yet.</p>

<p>Also much more important: <a href="https://issues.jenkins.io/browse/JENKINS-53547" title="Parallel sequential pipeline rendering appears distorted while jobs are running" class="issue-link" data-issue-key="JENKINS-53547">JENKINS-53547</a></p>

**Comment by jenkins-infra-bot** [comment]
<i>[Original `is_duplicated_by` from Jira: <a href="https://github.com/jenkinsci/blueocean-plugin/issues?q=is%3Aissue%20%22jira_issue_key%3DJENKINS-38442%22">JENKINS-38442</a>]</i>
